pkg/internal/kubectl/kubectl.go

```diff
@@ -14,91 +14,126 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+//Package kubectl executes various kubectl sub-commands in a forked shell
+//   OLD - mockgen -destination=pkg/internal/kubectl/mockKubectl.go -package=kubectl -source=pkg/internal/kubectl/kubectl.go gitlab.fmr.com/common-platform/addons-manager/pkg/internal/kubectl Kubectl,Command
+//go:generate mockgen -destination=mockKubectl.go -package=kubectl -source=kubectl.go . Kubectl,Command
 package kubectl
 
 import (
 	"fmt"
-	"os/exec"
 	"strings"
 
 	"github.com/go-logr/logr"
 )
 
-// Kubectl provides an api for creating and running kubectl sub-commands.
-type Kubectl struct {
-	logger  logr.Logger
-	cmdPath string
+var (
+	kubectlCmd   = "kubectl"
+	execProvider = NewExecProvider()
+)
+
+// Kubectl is a Factory interface that returns concrete Command implementations from named constructors
+type Kubectl interface {
+	Apply(path string) (c Command)
+	Delete(args ...string) (c Command)
+	Get(args ...string) (c Command)
+	getLogger() logr.Logger
+	getPath() string
+}
+
+// CommandFactory is a concrete Factory implementation of the Kubectl interface's API
+type CommandFactory struct {
+	logger logr.Logger
+	path   string
 }
 
-// NewKubectl returns a Kubectl object for creating and running kubectl sub-commands.
-func NewKubectl(l logr.Logger) (kctl *Kubectl, err error) {
-	kctl = &Kubectl{
-		logger: l,
+// NewKubectl returns a Kubectl object for creating and running kubectl sub-commands
+func NewKubectl(logger logr.Logger) (factory Kubectl, err error) {
```

Comment:
Why is this needed? kubectl shouldn't be used from inside a controller. client-go should give you everything you need.

Reply:
@richardcase - In this case, client-go does not give us everything we need.  Most of the logic that goes into applying YAML to the cluster lives in the `kubectl` code and is not server-side or broken out into a module that other projects can pick up.  If you look into similar open-source projects, they generally shell out to `kubectl` rather than re-implement all that logic themselves, and there's no other way currently to guarantee full compatibility.

I was also specifically directed to use `kubectl` in this manner.